src/model/MutableMediaDataSource.ts

```diff
@@ -1,119 +1,143 @@
-        const filter = { mediaUuid: doc.mediaUuid, destinationId }
+        const doc: EntityTag = {
+          _id: new mongoose.Types.ObjectId(),
+          targetId: entityUuid,
+          type: entityType,
+          ancestors: climb.parent.ancestors,
+          climbName: climb.name,
+          areaName: climb.parent.area_name,
+          lnglat: climb.metadata.lnglat
+        }
 
-        if (await media.exists(filter) != null) throw new UserInputError('Duplicate mediaUuid and destinationId not allowed')
+        // We treat 'entityTags' like a Set - can't tag the same climb/area id twice.
+        // See https://stackoverflow.com/questions/33576223/using-mongoose-mongodb-addtoset-functionality-on-array-of-objects
+        const filter = {
+          _id: new mongoose.Types.ObjectId(mediaId),
+          'entityTags.targetId': { $ne: entityUuid }
+        }
 
-        const rs = await media
-          .findOneAndUpdate(
+        await this.mediaObjectModel
+          .updateOne(
             filter,
-            doc,
-            QUERY_OPTIONS)
+            {
+              $push: {
+                entityTags: doc
+              }
+            })
+          .orFail(new UserInputError('Media not found or tag already exists.'))
           .lean()
 
-        if (rs == null) return rs
-
-        const climbTag: ClimbTagType = {
-          _id: rs._id,
-          mediaUuid: rs.mediaUuid,
-          mediaType: rs.mediaType,
-          mediaUrl: rs.mediaUrl,
-          destType: rs.destType,
-          climb: climb as ClimbType,
-          onModel: RefModelType.climbs
-        }
-        return climbTag
+        return doc
       }
 
       case 1: {
-        modelType = RefModelType.areas
         // Check whether the area referencing this tag exists before we allow
         // the tag to be added
-        const area = await getAreaModel()
-          .findOne({ 'metadata.area_id': destinationId })
-          .lean()
-          .orFail(new UserInputError(`Area with id: ${destinationId.toUUID().toString()} doesn't exist`))
-
-        const doc: MediaType = {
-          mediaUuid,
-          mediaType,
-          mediaUrl,
-          destType,
-          destinationId,
-          onModel: modelType
+        const area = await this.areaDS.findOneAreaByUUID(entityUuid)
+
+        if (area == null) {
+          throw new UserInputError(`Area with id: ${entityUuid.toUUID().toString()} not found`)
         }
 
-        const filter = { mediaUuid: doc.mediaUuid, destinationId }
+        const doc: EntityTag = {
+          _id: new mongoose.Types.ObjectId(),
+          targetId: entityUuid,
+          type: entityType,
+          ancestors: area.ancestors,
+          areaName: area.area_name,
+          lnglat: area.metadata.lnglat
+        }
 
-        if (await media.exists(filter) != null) throw new UserInputError('Duplicate mediaUuid and destinationId now allowed')
+        // We treat 'entityTags' like a Set - can't tag the same climb/area id twice.
```

Comment:
Small thing but this part onward is duplicated in case 0 above. Might be worth moving it out of the switch statement to deduplicate.

Reply:
Good point.  I'll submit a change.